### PR TITLE
Add contract definitions

### PR DESCRIPTION
**Pull request by joshbwlng**
Change-type: patch
Signed-off-by: Josh Bowling <josh@balena.io>

### CODE REVIEW

File: contracts.yml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+handle: external-event
```
Comment (joshbwlng) [inline contracts.yml:1]: @codewithcheese What happens when this type already exists in Jellyfish?

Comment (codewithcheese) [reply contracts.yml:1]: @joshbwlng for now there is not check, but we will brainstorm a way to reference and existing type or to denote that the definition is a reference to an exisiting type

**Comment by joshbwlng** [comment]
@codewithcheese Getting this strange type output after running `npx transformer-sdk bundle generate` (v0.3.1) in this branch:
```
export interface ExternalEventData {
	source: &#39;sentry&#39;;
	[k: string]: unknown;
}
```

It has to do with trying to match on `data.source="sentry"`.

**Comment by codewithcheese** [comment]
@joshbwlng that was caused by escaping in the template, i have disabled escaping for that block https://github.com/product-os/transformer-sdk/pull/10

**Comment by joshbwlng** [comment]
@codewithcheese Something seems to be going wrong with type generation? TypeScript is complaining that required base contract fields such as `created_at` are missing. Any ideas? (Also had to `npm i -D @balena/jellyfish-types` to satisfy missing dependencies, but that didn't solve this problem.)

type-issue
external-event

**Comment by codewithcheese** [comment]
@joshbwlng oh, `InputManifest` expects `Contract` as input, `Contract` is a full formed JF contract https://github.com/product-os/transformer-sdk/pull/11

Sorry you've been my guinea pig, hope its not to painful.

**Comment by joshbwlng** [comment]
@codewithcheese No worries, happy to provide useful feedback

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        slug: `diagnostic-event-${uuidv4()}`,
+        data: {
+            fingerprint: input.input.contract.data.payload.fingerprint,
+            priority: input.input.contract.data.payload.priority,
+            message: input.input.contract.data.payload.message,
+            timestamp: input.input.contract.data.payload.timestamp,
+        },
+        artifact: {
+            '$export': 'foobar',
+        },
```
Comment (joshbwlng) [inline index.ts]: @codewithcheese What should this be?

Comment (codewithcheese) [reply index.ts]: @joshbwlng what was your motivation for adding it? The artifact field is not required as output

Comment (joshbwlng) [reply index.ts]: @codewithcheese Getting a type error, claiming that it is required:
type-error

Comment (codewithcheese) [reply index.ts]: @joshbwlng ahh my bad, https://github.com/product-os/transformer-sdk/pull/12

File: contracts.yml [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+        - type
+        - data
+      properties:
+        type: { const: external-event }
+        data:
+          type: object
+          required:
+            - payload
+            - source
+          properties:
```
Comment (codewithcheese) [inline contracts.yml:53]: @joshbwlng is filtering on these properties to have a specific type necessary? Can you presume their type is correct given the contract type?

**Comment by joshbwlng** [comment]
@codewithcheese Are you referring to `data.payload.fingerprint` etc in the `external-event` contract? Do you think it would be better to only assert `data.source="sentry"` and then let the transformer function deal with payload properties?

Also, I'm trying to write a unit test for the transformer function and not sure what to provide for `input.transformerContract` and `input.artifactPath`. Apologies for all of the questions and thanks in advance!
function-types

**Comment by codewithcheese** [comment]
> Are you referring to data.payload.fingerprint etc in the external-event contract? Do you think it would be better to only assert data.source="sentry" and then let the transformer function deal with payload properties?

Sorry the comment didn't maintain the context I wanted, I changed it to a review.

**Comment by codewithcheese** [comment]
> Also, I'm trying to write a unit test for the transformer function and not sure what to provide for input.transformerContract and input.artifactPath. Apologies for all of the questions and thanks in advance!

Hmm we need good ways to mock this stuff, `transformerContract` would be the full transformer contract from JF. We could read the transformer contract in `contracts.yml`, and fake the other values.

The artifactPath can be any path unless your transformer works with artifacts in which case it should be the path to test artifact (fixture)

**Comment by codewithcheese** [comment]
All questions very welcome

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+	ExternalEventContract,
+	DiagnosticEventContractDefinition,
+} from './types';
+
+export async function externalevent2diagnosticevent(
+	input: InputManifest<ExternalEventContract>,
+): Promise<Array<Result<DiagnosticEventContractDefinition>>> {
+	const handle = 'diagnostic-event';
+	const diagnosticEvent: DiagnosticEventContractDefinition = {
+		handle,
```
Comment (joshbwlng) [inline index.ts]: @codewithcheese Is `handle` a new required field for contracts here?

Comment (codewithcheese) [reply index.ts]: @joshbwlng yeah I set it to required. It's part of transformers stage 3 to add a handle field to all contracts.

Comment (codewithcheese) [reply index.ts]: the handle is the machine readable name

Comment (codewithcheese) [reply index.ts]: @joshbwlng the handle is the identifier for this instance. It should not be the type as well. Unless the type was a singleton or something with only 1 instance for the type with a same handle.

File: index.spec.ts [hunk 5]
```diff
@@ -0,0 +1,57 @@
+import { externalevent2diagnosticevent } from './index';
+import type { ExternalEventContract } from './types';
+import { v4 as uuidv4 } from 'uuid';
+
+describe('Diagnostic event life-cycle', function () {
+	it('should transform', async function () {
+		const externalEvent: ExternalEventContract = {
+			type: 'external-event',
```
Comment (codewithcheese) [inline index.spec.ts:8]: @joshbwlng I updated the sdk to require a handle as input https://github.com/product-os/transformer-sdk/pull/14

What is the handle for an external event? I think that something we need to discuss, how is a external event slug generated now, perhaps some randomness is added? I cannot access them via the sdk

Comment (joshbwlng) [reply index.spec.ts:8]: @codewithcheese The `external-event` slug is generated here: https://github.com/product-os/jellyfish/blob/master/apps/server/lib/http/routes.ts#L586-L588

So just `external-event-${uuidv4()}`. It's still unclear to me why a handle is necessary if its only a identifier for an instance of a type when we already have id and slug? I'm sure I'm just missing something though.

Comment (joshbwlng) [reply index.spec.ts:8]: For now I'll just do the same for handles, type + uuidv4.

File: contracts.yml (hunk 3 above)
Comment (joshbwlng) [reply contracts.yml:53]: @codewithcheese Nope, you're right, I've reduced it down to what is minimally needed

File: index.spec.ts (hunk 5 above)
Comment (joshbwlng) [reply index.spec.ts:8]: Ok, from what I understand handles are scoped to the type and globally unique to all types. Also, the handle isn't necessarily generated by Jellyfish, but instead in control of the user/agent creating the contract. Let me know if this is wrong and/or missing something.

**Comment by joshbwlng** [comment]
@balena-ci I self certify!

**Comment by joshbwlng** [comment]
@balena-ci I self certify!

**Pull request merged**